### PR TITLE
refactor(swap): fetch all asset prices at once

### PR DESCRIPTION
**Pull request by acdibble**

### CODE REVIEW

File: packages/swap/src/pricing/index.ts [hunk 1]
```diff
@@ -1,58 +1,97 @@
...
+
+  shouldRefetch() {
+    return Date.now() - this.lastCacheSet < PriceCache.TTL;
   }
 
-  const { promise, resolve } = deferredPromise<number | undefined>();
+  async getAssetPrice(asset: InternalAsset): Promise<number | undefined> {
+    logger.debug(`getting asset price for "${asset}"`);
+
+    if (this.shouldRefetch()) {
```
Comment (jerryafr) [inline packages/swap/src/pricing/index.ts:45]: I think it should not be called shouldRefetch, but isUpToDate or something like that?

Comment (acdibble) [reply packages/swap/src/pricing/index.ts]: Yep, renamed

**Review by jerryafr — APPROVED** [review]

File: packages/swap/src/pricing/index.ts [hunk 2]
```diff
@@ -1,58 +1,97 @@
...
+
+  freshEnough() {
+    return Date.now() - this.lastCacheSet < PriceCache.TTL;
   }
 
-  const { promise, resolve } = deferredPromise<number | undefined>();
+  async getAssetPrice(asset: InternalAsset): Promise<number | undefined> {
+    logger.debug(`getting asset price for "${asset}"`);
+
+    if (this.freshEnough()) {
```
Comment (zoheb391) [inline packages/swap/src/pricing/index.ts:45]: image

File: packages/swap/src/pricing/index.ts [hunk 3]
```diff
@@ -1,58 +1,97 @@
 import axios from 'axios';
 import { InternalAsset } from '@/shared/enums';
 import env from '../config/env';
-import { CacheMap } from '../utils/dataStructures';
```
Comment (niklasnatter) [inline packages/swap/src/pricing/index.ts:4]: we removed something from the `CacheMap` to make it work for this usecase, so maybe we should readd that?
https://github.com/chainflip-io/chainflip-sdk-monorepo/pull/570/files#diff-b0ae927d6867edde94cb13c8bf6099e9f97f6e27905f178c4e481a70ec814e5dL12

Comment (acdibble) [reply packages/swap/src/pricing/index.ts:4]: done

File: packages/swap/src/pricing/index.ts [hunk 4]
```diff
@@ -1,58 +1,97 @@
...
 
 const coingeckoAxios = axios.create({
   baseURL: 'https://pro-api.coingecko.com/api/v3',
   timeout: 5000,
   headers: { 'x-cg-pro-api-key': env.COINGECKO_API_KEY },
 });
 
-export const getAssetPrice = async (asset: InternalAsset): Promise<number | undefined> => {
-  logger.debug(`getting asset price for "${asset}"`);
+export class PriceCache {
```
Comment (acdibble) [inline packages/swap/src/pricing/index.ts:29]: using a class for encapsulation is easier than trying to reset global variables by resetting mocks in jest

**Pull request merged**